### PR TITLE
Fix tfw_http_msg_write() (#210)

### PR DESCRIPTION
**Pull request by keshonok**

### CODE REVIEW

File: tempesta_fw/http_msg.c [hunk 1]
```diff
@@ -559,40 +559,58 @@ tfw_http_msg_iter_init(TfwHttpMsg *hm, TfwMsgIter *it)
...
 			frag = ss_skb_frag_next(&hm->msg.skb_list,
 						&it->skb, &it->frag);
+			it->frag_off = 0;
+			/*
+			 * If all data from the chunk has been copied,
+			 * then switch to next chunk. Otherwise, stay
+			 * in the current chunk.
+			 */
+			/* Note that n_copy equals f_room here. */
+			if (c_size == n_copy) {
```
Comment (krizhanovsky) [inline tempesta_fw/http_msg.c]: It'd be easier to understand if you replace the condition by ```c_size < f_room```

File: tempesta_fw/ss_skb.h [hunk 2]
```diff
@@ -167,6 +167,23 @@ ss_skb_adjust_data_len(struct sk_buff *skb, int delta)
...
+ * In that case, return PAGE_SIZE for the fragment size.
+ * See implementation of ss_skb_alloc().
+ */
+static inline unsigned int
+ss_skb_frag_size(skb_frag_t *frag, unsigned int frag_size)
+{
+        return frag_size ? skb_frag_size(frag) : PAGE_SIZE;
+}
+
+
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.h]: extra blank line

File: tempesta_fw/http_msg.c [hunk 3]
```diff
@@ -549,50 +549,109 @@ tfw_http_msg_hdr_add(TfwHttpMsg *hm, TfwStr *hdr)
...
+ * to support consecutive writes. This function is intended to work
+ * together with tfw_http_msg_write() that uses the @it iterator.
+ */
+TfwHttpMsg *
+tfw_http_msg_create(TfwMsgIter *it, int type, size_t data_len)
+{
+	TfwHttpMsg *hm = tfw_http_msg_alloc(type);
+
+	if ((hm == NULL) || (data_len == 0))
+		return NULL;
```
Comment (krizhanovsky) [inline tempesta_fw/http_msg.c]: It seems we should return either @hm if data_len == 0 (@hm is still valid new message, but with zero body) or free the message before return. I'd prefer the first one.

File: tempesta_fw/http_msg.c [hunk 4]
```diff
@@ -549,50 +549,109 @@ tfw_http_msg_hdr_add(TfwHttpMsg *hm, TfwStr *hdr)
...
+/*
+ * Fill up an HTTP message @hm with data from string @data. An iterator
+ * @it is used to support multiple calls to this functions after set up.
+ * This function can only be called after a call to tfw_http_msg_setup().
+ * It works only with empty SKB space prepared by tfw_http_msg_setup().
+ * It should not be used under any other circumstances.
+ */
 int
-tfw_http_msg_write(TfwHttpMsg *hm, TfwMsgIter *it, const TfwStr *data)
+tfw_http_msg_write(TfwMsgIter *it, TfwHttpMsg *hm, const TfwStr *data)
```
Comment (krizhanovsky) [inline tempesta_fw/http_msg.c]: Not a big deal, but usually return values are the last one.

File: tempesta_fw/http_msg.h [hunk 5]
```diff
@@ -24,8 +24,9 @@
 #include "http.h"
 
 typedef struct {
-	int		frag;
-	int		frag_off;
+	unsigned int	frag;
+	unsigned int	frag_off;
+	unsigned int	frag_size;
```
Comment (krizhanovsky) [inline tempesta_fw/http_msg.h]: It seem you forgot to remove frag_size.

File: tempesta_fw/http_msg.c (hunk 3 above)
Comment (keshonok) [inline tempesta_fw/http_msg.c]: On return from this function we expect that it has successfully created both the message and all SKBs with data. So it has to be all or nothing here.

**Pull request merged**